pkg/pubsubsource/pubsub.go

```diff
@@ -0,0 +1,101 @@
+/*
+Copyright 2022 The Numaproj Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package pubsubsource
+
+import (
+	"context"
+	"log"
+	"sync"
+	"time"
+
+	"cloud.google.com/go/pubsub"
+	sourcesdk "github.com/numaproj/numaflow-go/pkg/sourcer"
+)
+
+// PubSubSource represents a source of messages in a publish-subscribe system.
+type PubSubSource struct {
+	client       *pubsub.Client
+	subscription *pubsub.Subscription
+	receiveCh    chan *pubsub.Message       //buffered channel for buffering messages received from the subscription.
+	messages     map[string]*pubsub.Message //messages field  serves as a cache or storage for unacknowledged pubsub.Message objects that have been received but not yet processed.
+	lock         *sync.Mutex
+}
+
+func NewPubSubSource(client *pubsub.Client, subscription *pubsub.Subscription, maxExtensionPeriod time.Duration, maxOutStandingMessages int) *PubSubSource {
+	subscription.ReceiveSettings.MaxExtension = maxExtensionPeriod
+	subscription.ReceiveSettings.MaxOutstandingMessages = maxOutStandingMessages
+	receiveCh := make(chan *pubsub.Message, maxOutStandingMessages)
+	return &PubSubSource{client: client, subscription: subscription, receiveCh: receiveCh, messages: make(map[string]*pubsub.Message), lock: new(sync.Mutex)}
+}
+
+func (p *PubSubSource) Read(_ context.Context, readRequest sourcesdk.ReadRequest, messageCh chan<- sourcesdk.Message) {
+
+	// this condition will happen only if there are unacked messages.
+	// Read can happen only after all the messages are acked.
+	if len(p.messages) > 0 {
+		return
+	}
+	ctx, cancel := context.WithTimeout(context.Background(), readRequest.TimeOut())
+	defer cancel()
+	for i := 0; i < int(readRequest.Count()) && ctx.Err() == nil; i++ {
```

Comment:
Why do we need ctx.Err() == nil?

Reply:
just for safety here ,we are already relying on select for ctx.Done ,but for double check,can be removed though if you recommend @KeranYang

Reply:
I've seen quite a lot select with ctx.Done, never saw one with this extra check. Let's remove it.